### PR TITLE
Android Gradle Plugin 3.1, Gradle 4.3, Java 8, SDK 27, Library Updates, Badges, New D8 Compiler Enabled, Documentation Improvements

### PR DESCRIPTION
**Pull request by vivekkiran**
https://android-developers.googleblog.com/2017/08/next-generation-dex-compiler-now-in.html
https://developer.android.com/topic/libraries/support-library/revisions.html
https://developer.android.com/studio/releases/index.html
https://developer.android.com/studio/preview/index.html
https://www.infoq.com/news/2015/05/Oracle-Ends-Java-7Public-Updates
https://developer.ibm.com/wasdev/docs/end-java-se-6-go/
https://shields.io/
https://github.com/gradle/gradle/releases

### CODE REVIEW

File: realm-transformer/gradle.properties [hunk 1]
```diff
@@ -1 +1,2 @@
 org.gradle.caching=true
+android.enableD8=true
```
Comment (zaki50) [inline realm-transformer/gradle.properties]: No need to add this.
`realm-transformer` never be converted to dex.

File: realm/gradle.properties [hunk 2]
```diff
@@ -1,2 +1,3 @@
 org.gradle.jvmargs=-Xms512m -Xmx2048m
 org.gradle.caching=true
+android.enableD8=true
```
Comment (zaki50) [inline realm/gradle.properties]: No need to add this.
D8 doesn't affect library project.

File: realm-annotations/gradle.properties [hunk 3]
```diff
@@ -1 +1,2 @@
 org.gradle.caching=true
+android.enableD8=true
```
Comment (zaki50) [inline realm-annotations/gradle.properties]: No need to add this.
`realm-annotations` is pure Java project.

File: gradle.properties [hunk 4]
```diff
@@ -1,2 +1,3 @@
 org.gradle.jvmargs=-XX:MaxPermSize=512m
 org.gradle.caching=true
+android.enableD8=true
```
Comment (zaki50) [inline gradle.properties]: No need to add this.

File: gradle-plugin/gradle.properties [hunk 5]
```diff
@@ -1 +1,2 @@
 org.gradle.caching=true
+android.enableD8=true
```
Comment (zaki50) [inline gradle-plugin/gradle.properties]: No need to add this.
`realm-gradle` is a pure Java project.

File: examples/gradle.properties [hunk 6]
```diff
@@ -1,5 +1,6 @@
 org.gradle.jvmargs=-Xmx2048M
 org.gradle.caching=true
+android.enableD8=true
```
Comment (zaki50) [inline examples/gradle.properties]: You should remove `examples/newsreaderExample/gradle.properties` if you enable D8 in all example apps.

File: gradle.properties (hunk 4 above)
Comment (vivekkiran) [reply gradle.properties]: I have removed d8 from the library modules 👍

File: examples/gradle.properties (hunk 6 above)
Comment (vivekkiran) [reply examples/gradle.properties]: 👍

**Comment by zaki50** [comment]
@vivekkiran We have `tools/update_gradle_wrapper.sh` script which regenerates gradle wrappers.
After modifying `realm.properties`, could you execute it in order to keep gradle wrappers up to date?

**Comment by vivekkiran** [comment]
@zaki50 regenerated the wrappers :)

**Comment by vivekkiran** [comment]
screen shot 2017-10-26 at 8 59 32 pm

**Comment by zaki50** [comment]
@vivekkiran I think that we should use stable version of tools (e.g. AGP 3.0 and Gradle 4.2.1) if  there is strong demand for alpha/beta tools.

**Comment by zaki50** [comment]
@vivekkiran One more thing. Could you change target branch to `releases` instead of `master`?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,16 @@
+## 4.2.0 (YYYY-MM-DD)
```
Comment (beeender) [inline CHANGELOG.md]: 4.2.0 entry need to be removed since this is going to be merged to the `releases` branch.

File: examples/build.gradle [hunk 8]
```diff
@@ -32,7 +33,8 @@ allprojects {
             maven { url 'https://jitpack.io' }
         }
         dependencies {
-            classpath 'com.android.tools.build:gradle:3.0.0-rc2'
+            classpath 'com.android.tools.build:gradle:3.1.0-alpha01'
+            classpath 'com.novoda:gradle-android-command-plugin:1.7.1'
```
Comment (beeender) [inline examples/build.gradle]: this has been removed recently.

**Comment by beeender** [comment]
@vivekkiran I created #5506 to trigger CI for this, that contains all of you commits.

Would you please accept the CLA here so we can merge your commit? Follow the instructions here https://github.com/realm/realm-java/blob/master/CONTRIBUTING.md#cla

**Comment by vivekkiran** [comment]
@beeender I have accepted the CLA :)

File: version.txt [hunk 9]
```diff
@@ -1 +1 @@
-4.1.2-SNAPSHOT
\ No newline at end of file
+4.2.0-SNAPSHOT
```
[the remark this answers is not in the document]
Comment (beeender) [reply version.txt]: i fixed this in #5506

**Comment by beeender** [comment]
@vivekkiran i will address @nhachicha 's comments in #5472 and merge it.

**Comment by vivekkiran** [comment]
@nhachicha @beeender go ahead :) 👍

File: README.md [hunk 10]
```diff
@@ -21,18 +24,20 @@ The API reference is located at [realm.io/docs/java/api](https://realm.io/docs/j
...
 - **Have a feature request?** [Open an issue](https://github.com/realm/realm-java/issues/new). Tell us what the feature should do, and why you want the feature.
 - Sign up for our [**Community Newsletter**](https://go.pardot.com/l/210132/2017-04-26/3j74l) to get regular tips, learn about other use-cases and get alerted of blogposts and tutorials about Realm.
 
 ## Using Snapshots
 
-If you want to test recent bugfixes or features that have not been packaged in an official release yet, you can use a **-SNAPSHOT** release of the current development version of Realm via Gradle, available on [Jfrog OSS](http://oss.jfrog.org/oss-snapshot-local/io/realm/realm-gradle-plugin/)
+If you want to test recent bugfixes or features that have not been packaged in an official release yet, you can use a **-SNAPSHOT** release of the current development version of Realm via Gradle, available on [JFrog OSS](http://oss.jfrog.org/oss-snapshot-local/io/realm/realm-gradle-plugin/)
 
-```gradle
+```In build.gradle
```
[the remark this answers is not in the document]
Comment (beeender) [reply README.md]: Fixed, should be `groovy`

**Comment by beeender** [comment]
#5472 will be merged instead.

**Pull request closed without merging**